src/serializationStrategies/XpathRangeSelector/xpath.js

```diff
@@ -1,31 +1,140 @@
-// mostly copied from https://developer.mozilla.org/en-US/docs/Web/XPath/Snippets#getXPathForElement
-export function getXPathForElement(el, reference) {
+import {DATA_ATTR} from '../../injectHighlightWrappers';
+
+const findElementChild = node => Array.prototype.find.call(node.childNodes, node => node.nodeType === 1);
+const isHighlight = node => node && node.getAttribute && node.getAttribute(DATA_ATTR) !== null;
+const isTextHighlight = node => isHighlight(node) && !findElementChild(node);
+const isText = node => node && node.nodeType === 3;
+const isElement = node => node && node.nodeType === 1;
+
+// kinda copied from https://developer.mozilla.org/en-US/docs/Web/XPath/Snippets#getXPathForElement
+export function getXPathForElement(targetElement, offset, reference) {
   var xpath = '';
-  var pos, tempitem2;
-
-  while(el !== reference) {
-    pos = 0;
-    tempitem2 = el;
-    while(tempitem2) {
-      if ([3, 1].indexOf(tempitem2.nodeType) !== -1 && tempitem2.nodeName === el.nodeName) { // If it is ELEMENT_NODE or TEXT_NODE of the same name
-        pos += 1;
+  var pos
+    , element = targetElement.previousSibling
+    , focus = targetElement;
+
+  // for a text target, highlights might have broken up the text node,
+  // look for preceeding nodes that need to be combined into this one
+  // and modify the range offset accordingly
+  if (isText(focus) && isTextHighlight(element)) {
+    while(element && (isText(element) || isTextHighlight(element))) {
+      offset += element.textContent.length;
+      element = element.previousSibling;
+    }
+  // for element targets, highlight children might be artifically
+  // inflating the range offset, fix.
+  } else if (isElement(focus)) {
+    offset -= Array.prototype.filter.call(Array.prototype.slice.call(focus.childNodes, 0, offset), isHighlight).length;
+  }
+
+  while(focus !== reference) {
+    pos = 1;
+
+    while(element) {
+      // highlights in text change the number of nodes in the nodelist,
+      // compensate by gobbling adjacent highlights and text
+      if (isText(focus) && isTextHighlight(element)) {
+        while(element && (isText(element) || isTextHighlight(element))) {
+          element = element.previousSibling;
+        }
+      }
+      if (element) {
+        if ([3, 1].indexOf(element.nodeType) !== -1 && !isHighlight(element) && element.nodeName === focus.nodeName) { // If it is ELEMENT_NODE or TEXT_NODE of the same name
+          pos += 1;
```

Comment:
since we're not supporting IE, could `use [3,1].includes(element.nodeType)` or even better add another helper func `isElmentOrText` and use that

Reply:
added the `isTextOrElement` helper